packages/core-contracts/contracts/proxy/UUPSImplementation.sol

```diff
@@ -5,23 +5,75 @@ import "../utils/ContractUtil.sol";
 import "../common/CommonErrors.sol";
 import "./ProxyStorage.sol";
 
-contract UUPSImplementation is ProxyStorage, ContractUtil, CommonErrors {
+abstract contract UUPSImplementation is ProxyStorage, ContractUtil, CommonErrors {
+    error SterileImplementation(address implementation);
+    error SimulatedUpgradeFailed();
+    error UpgradeToNotCalledViaProxy();
+
     event Upgraded(address implementation);
 
-    function _upgradeTo(address newImplementation) internal virtual {
+    // WARNING!!!
+    // **************************************************************************
+    // This function is intentionally not implemented, because
+    // it is critical and should not be exposed by default.
+    // It should be implemented with some protection like `onlyOwner`,
+    // and simply call the underlying intenal function.
+    // **************************************************************************
+    function upgradeTo(address newImplementation) public virtual;
+
```

Comment:
Below is some black magic to avoid the state boolean that was needed to avoid recursion in the fertility check.